File: app/src/main/java/com/antyzero/smoksmog/ui/IndicatorView.java
package com.antyzero.smoksmog.ui;

import android.animation.ValueAnimator;
import android.annotation.TargetApi;
import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.RectF;
import android.graphics.Typeface;
import android.text.StaticLayout;
import android.text.TextPaint;
import android.util.AttributeSet;
import android.util.DisplayMetrics;
import android.util.TypedValue;
import android.view.View;
import android.view.animation.AccelerateDecelerateInterpolator;

import pl.malopolska.smoksmog.model.Particulate;

public class IndicatorView extends View {

    private static final float GAP_ANGLE_DEFAULT = 90;
    private static final int STROKE_WIDTH_DEFAULT = 10;
    public static final int STEP_COLOR = 50;

    private float gapAngle = GAP_ANGLE_DEFAULT;

    private final ValueAnimator valueAnimator = ValueAnimator.ofFloat( 0f );
    private final ValueAnimator valueMaxAnimator = ValueAnimator.ofFloat( 0f );

    private Paint paintArcBackground;
    private Paint paintArcForeground;
    private Paint paintNameShort;
    private Paint paintValue;
    private Paint paintMaxValue;

    private float startAngle;
    private float sweepAngle;
    private float strokeWidth;
    private float canvasHorizontalMiddle;
    private float canvasVerticalMiddle;

    private int currentWidth;
    private int currentHeight;

    private float value = 0f;
    private float valueMax = 100f;
    private float arcValue = 0f;
    private float textNameSize = 10f;

    private RectF arcRect;
    private Rect textShortNameRect;

    private String particulateName = "";

    private int overLap = 0;

    private int arcPositionOffset = 0;
    private String valueText = "";

    /**
     * Setup variables
     */ {
        valueAnimator.setDuration( 1000L );
        valueAnimator.setInterpolator( new AccelerateDecelerateInterpolator() );

        valueMaxAnimator.setDuration( 1000L );
        valueMaxAnimator.setInterpolator( new AccelerateDecelerateInterpolator() );
    }

    public IndicatorView( Context context ) {
        super( context );
        init( context );
    }

    public IndicatorView( Context context, AttributeSet attrs ) {
        super( context, attrs );
        init( context );
    }

    public IndicatorView( Context context, AttributeSet attrs, int defStyleAttr ) {
        super( context, attrs, defStyleAttr );
        init( context );
    }

    @SuppressWarnings( "unused" )
    @TargetApi( 21 )
    public IndicatorView( Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes ) {
        super( context, attrs, defStyleAttr, defStyleRes );
        init( context );
    }

    private void init( Context context ) {

        strokeWidth = dipToPixels( context, STROKE_WIDTH_DEFAULT );
        textNameSize = dipToPixels( context, 32 );

        // Init paints

        paintArcBackground = new Paint();
        paintArcBackground.setAntiAlias( true );
        paintArcBackground.setStrokeCap( Paint.Cap.ROUND );
        paintArcBackground.setStrokeWidth( strokeWidth - 2 );
        paintArcBackground.setStyle( Paint.Style.STROKE );

        paintArcForeground = new Paint( paintArcBackground );
        paintArcForeground.setAntiAlias( true );
        paintArcForeground.setStrokeCap( Paint.Cap.ROUND );
        paintArcForeground.setStyle( Paint.Style.STROKE );
        paintArcForeground.setColor( Color.CYAN );
        paintArcForeground.setStrokeWidth( strokeWidth );

        paintValue = new Paint();
        paintValue.setAntiAlias( true );
        paintValue.setColor( Color.RED );
        paintValue.setStyle( Paint.Style.FILL );
        paintValue.setTypeface( Typeface.defaultFromStyle( Typeface.BOLD ) );

        paintMaxValue = new Paint();
        paintMaxValue.setAntiAlias( true );
        paintMaxValue.setColor( Color.RED );
        paintMaxValue.setStyle( Paint.Style.FILL );
        paintMaxValue.setTypeface( Typeface.defaultFromStyle( Typeface.BOLD ) );

        paintNameShort = new Paint();
        paintNameShort.setAntiAlias( true );
        paintNameShort.setColor( Color.RED );
        paintNameShort.setStyle( Paint.Style.FILL );
        paintNameShort.setTextSize( textNameSize );

        calculateMinorDrawingVariables();
    }

    @Override
    protected void onMeasure( int widthMeasureSpec, int heightMeasureSpec ) {
        int widthMode = MeasureSpec.getMode( widthMeasureSpec );
        int widthSize = MeasureSpec.getSize( widthMeasureSpec );
        int heightMode = MeasureSpec.getMode( heightMeasureSpec );
        int heightSize = MeasureSpec.getSize( heightMeasureSpec );

        int size;
        if ( widthMode == MeasureSpec.EXACTLY && widthSize > 0 ) {
            size = widthSize;
        } else if ( heightMode == MeasureSpec.EXACTLY && heightSize > 0 ) {
            size = heightSize;
        } else {
            size = widthSize < heightSize ? widthSize : heightSize;
        }

        int widthNewSpec = MeasureSpec.makeMeasureSpec( size, MeasureSpec.EXACTLY );
        int heightNewSpec = MeasureSpec.makeMeasureSpec( size * 2, MeasureSpec.EXACTLY );
        super.onMeasure( widthNewSpec, heightNewSpec );
    }

    @Override
    protected void onSizeChanged( int w, int h, int oldw, int oldh ) {
        super.onSizeChanged( w, h, oldw, oldh );
        currentWidth = w;
        currentHeight = h;
        calculateDrawingVariables();
    }

    private static float dipToPixels( Context context, float dipValue ) {
        DisplayMetrics metrics = context.getResources().getDisplayMetrics();
        return TypedValue.applyDimension( TypedValue.COMPLEX_UNIT_DIP, dipValue, metrics );
    }

    public void setParticulate( Particulate particulate ) {
        setValueMax( particulate.getNorm() );
        setValue( particulate.getValue() );
        this.particulateName = particulate.getShortName();
    }

    private void setValue( float newValue ) {

        this.overLap = ( int ) ( value / valueMax ); // How many x times more that max

        valueAnimator.end();
        valueAnimator.setFloatValues( this.value, newValue );
        valueAnimator.addUpdateListener( animation -> {
            this.value = ( float ) animation.getAnimatedValue();
            recalculateDuringAnimation();
        } );
        valueAnimator.start();
    }

    /**
     * Value considered to be max value or call it 'norm' for this indicator view
     *
     * @param newValueMax to use
     */
    private void setValueMax( float newValueMax ) {
        this.valueMax = newValueMax;
    }

    /**
     * This should be called when animating values
     */
    private void recalculateDuringAnimation() {
        float progress = ( value % valueMax ) / valueMax;
        this.arcValue = progress * sweepAngle;
        this.valueText = String.format( "%.0f", value );
        this.overLap = ( int ) (value / valueMax);
        int red = ( int ) Math.min( STEP_COLOR * overLap + STEP_COLOR * progress, 255 );
        this.paintArcBackground.setARGB( 255,red,0,0 );
        postInvalidateOnAnimation();
    }

    /**
     * Non ui dependant variables
     */
    private void calculateMinorDrawingVariables() {
        startAngle = 90f + gapAngle / 2;
        sweepAngle = 360 - gapAngle;
    }

    /**
     * Recalculate some variables, important when view changes its size
     */
    private void calculateDrawingVariables() {
        calculateMinorDrawingVariables();

        // Calculate areas

        arcPositionOffset = ( currentHeight - currentWidth ) / 2;

        textShortNameRect = new Rect( 0,0, currentWidth, arcPositionOffset );

        // ...

        final float halfStrokeWidth = strokeWidth / 2;

        //noinspection SuspiciousNameCombination
        arcRect = new RectF(
                halfStrokeWidth,
                halfStrokeWidth + arcPositionOffset,
                currentWidth - halfStrokeWidth,
                currentWidth - halfStrokeWidth + arcPositionOffset );

        canvasHorizontalMiddle = currentWidth / 2;
        canvasVerticalMiddle = currentWidth / 2;

        postInvalidateOnAnimation();
    }

    @Override
    protected void onDraw( Canvas canvas ) {
        super.onDraw( canvas );

        canvas.drawArc( arcRect, startAngle, sweepAngle, false, paintArcBackground );

        // TODO if overlap is > 0 draw pre background ?

        canvas.drawArc( arcRect, startAngle, arcValue, false, paintArcForeground );

        drawTextCentred( canvas, paintValue, valueText, arcRect );

        //drawTextCentred( canvas, paintMaxValue, normText, canvasHorizontalMiddle, arcRect.bottom - 150 );
        //drawTextCentred( canvas, paintNameShort, particulateName, canvasHorizontalMiddle, arcRect.bottom );
    }

    private void drawTextCentred( Canvas canvas, Paint paint, String text, RectF bounds ) {
        drawTextCentred( canvas, paint, text, bounds.centerX(), bounds.centerY(), new Rect() );
    }

    private void drawTextCentred( Canvas canvas, Paint paint, String text, float cx, float cy ) {
        drawTextCentred( canvas, paint, text, cx, cy, new Rect() );
    }

    private void drawTextCentred( Canvas canvas, Paint paint, String text, float cx, float cy, Rect textBounds ) {
        paint.getTextBounds( text, 0, text.length(), textBounds );
        canvas.drawText( text, cx - textBounds.exactCenterX(), cy - textBounds.exactCenterY(), paint );
    }
}

File: android/src/main/kotlin/smoksmog/ui/IndicatorView.kt
package smoksmog.ui

import android.animation.Animator
import android.animation.ValueAnimator
import android.annotation.TargetApi
import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect
import android.graphics.RectF
import android.graphics.Typeface
import android.util.AttributeSet
import android.util.DisplayMetrics
import android.util.TypedValue
import android.view.View
import android.view.animation.AccelerateDecelerateInterpolator

import smoksmog.R

@Suppress("DEPRECATION")
class IndicatorView : View {

    private lateinit var paintValue: Paint
    private lateinit var paintArcBackground: Paint
    private lateinit var paintArcForeground: Paint

    private val valueAnimator = ValueAnimator.ofFloat(0f)
    private val gapAngle = GAP_ANGLE_DEFAULT
    private var paintNameShort: Paint? = null
    private var paintMaxValue: Paint? = null

    private var startAngle: Float = 0.toFloat()
    private var sweepAngle: Float = 0.toFloat()
    private var strokeWidth: Float = 0.toFloat()
    private var canvasHorizontalMiddle: Float = 0.toFloat()
    private var canvasVerticalMiddle: Float = 0.toFloat()

    private var currentWidth: Int = 0
    private var currentHeight: Int = 0

    private var value = 0f
    private var arcValue = 0f
    private var textNameSize = 10f

    private var arcRect = RectF()

    private var overLap = 0

    private var valueText = ""
    private var colorGood: Int = 0
    private var colorBad: Int = 0

    /**
     * Setup variables
     */
    init {
        valueAnimator.duration = 1000L
        valueAnimator.interpolator = AccelerateDecelerateInterpolator()
    }

    constructor(context: Context) : super(context) {
        init(context)
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init(context)
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init(context)
    }

    @TargetApi(21)
    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes) {
        init(context)
    }

    private fun init(context: Context) {

        strokeWidth = dipToPixels(context, STROKE_WIDTH_DEFAULT.toFloat())
        textNameSize = dipToPixels(context, 32f)

        colorBad = context.resources.getColor(R.color.indicator_bad)
        colorGood = context.resources.getColor(R.color.indicator_good)

        // Init paints

        paintArcBackground = Paint()
        paintArcBackground.isAntiAlias = true
        paintArcBackground.strokeCap = Paint.Cap.ROUND
        paintArcBackground.strokeWidth = strokeWidth - 2
        paintArcBackground.style = Paint.Style.STROKE
        paintArcBackground.color = resources.getColor(R.color.iron)
        paintArcBackground.alpha = 104

        paintArcForeground = Paint(paintArcBackground)
        paintArcForeground.isAntiAlias = true
        paintArcForeground.strokeCap = Paint.Cap.ROUND
        paintArcForeground.style = Paint.Style.STROKE
        paintArcForeground.color = Color.parseColor("#5be6dc")
        paintArcForeground.strokeWidth = strokeWidth

        paintValue = Paint()
        paintValue.isAntiAlias = true
        paintValue.color = context.resources.getColor(R.color.iron)
        paintValue.style = Paint.Style.FILL
        paintValue.typeface = Typeface.createFromAsset(context.assets, "fonts/Lato-Light.ttf")

        paintMaxValue = Paint()
        paintMaxValue!!.isAntiAlias = true
        paintMaxValue!!.style = Paint.Style.FILL
        paintMaxValue!!.typeface = Typeface.defaultFromStyle(Typeface.BOLD)

        paintNameShort = Paint()
        paintNameShort!!.isAntiAlias = true
        paintNameShort!!.style = Paint.Style.FILL
        paintNameShort!!.textSize = textNameSize

        calculateMinorDrawingVariables()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val widthMode = View.MeasureSpec.getMode(widthMeasureSpec)
        val widthSize = View.MeasureSpec.getSize(widthMeasureSpec)
        val heightMode = View.MeasureSpec.getMode(heightMeasureSpec)
        val heightSize = View.MeasureSpec.getSize(heightMeasureSpec)

        val size: Int
        if (widthMode == View.MeasureSpec.EXACTLY && widthSize > 0) {
            size = widthSize
        } else if (heightMode == View.MeasureSpec.EXACTLY && heightSize > 0) {
            size = heightSize
        } else {
            size = if (widthSize < heightSize) widthSize else heightSize
        }

        val widthNewSpec = View.MeasureSpec.makeMeasureSpec(size, View.MeasureSpec.EXACTLY)
        val heightNewSpec = View.MeasureSpec.makeMeasureSpec(size, View.MeasureSpec.EXACTLY)
        super.onMeasure(widthNewSpec, heightNewSpec)
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        currentWidth = w
        currentHeight = h
        calculateDrawingVariables()
    }

    /**
     * As 0% -> 100% -> more

     * @param newValue
     */
    fun setValue(newValue: Float) {

        if (valueAnimator.isRunning) {
            valueAnimator.cancel()
        }

        valueAnimator.setFloatValues(this.value, newValue)
        valueAnimator.addUpdateListener { animation ->
            this@IndicatorView.value = animation.animatedValue as Float
            recalculateDuringAnimation()
        }
        valueAnimator.addListener(object : Animator.AnimatorListener {
            override fun onAnimationStart(animation: Animator) {
                // Do nothing
            }

            override fun onAnimationEnd(animation: Animator) {
                // Do nothing
            }

            override fun onAnimationCancel(animation: Animator) {
                if (animation is ValueAnimator) {
                    this@IndicatorView.value = animation.animatedValue as Float
                }
            }

            override fun onAnimationRepeat(animation: Animator) {
                // Do nothing
            }
        })
        valueAnimator.start()
    }

    /**
     * This should be called when animating values
     */
    private fun recalculateDuringAnimation() {

        this.arcValue = value * sweepAngle
        this.valueText = String.format("%.0f%%", value * 100)
        this.overLap = Math.floor(value.toDouble()).toInt()

        paintArcForeground.color = if (arcValue <= 360f)
            colorGood
        else
            Color.rgb(242, 162, 60)

        postInvalidateOnAnimation()
    }

    /**
     * Non ui dependant variables
     */
    private fun calculateMinorDrawingVariables() {
        startAngle = 270f + gapAngle / 2
        sweepAngle = 360 - gapAngle
    }

    /**
     * Recalculate some variables, important when view changes its size
     */
    private fun calculateDrawingVariables() {
        calculateMinorDrawingVariables()

        val halfStrokeWidth = strokeWidth / 2


        arcRect = RectF(
                halfStrokeWidth,
                halfStrokeWidth,
                currentWidth - halfStrokeWidth,
                currentWidth - halfStrokeWidth)

        canvasHorizontalMiddle = (currentWidth / 2).toFloat()
        canvasVerticalMiddle = (currentWidth / 2).toFloat()

        paintValue.textSize = (arcRect.bottom - arcRect.top) / 4f

        postInvalidateOnAnimation()
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        canvas.drawArc(arcRect, startAngle, sweepAngle, false, paintArcBackground)

        // TODO if overlap is > 0 draw pre background ?

        canvas.drawArc(arcRect, startAngle, arcValue, false, paintArcForeground)
        drawTextCentred(canvas, paintValue, valueText, arcRect)
    }

    private fun drawTextCentred(canvas: Canvas, paint: Paint, text: String, bounds: RectF) {
        drawTextCentred(canvas, paint, text, bounds.centerX(), bounds.centerY(), Rect())
    }

    private fun drawTextCentred(canvas: Canvas, paint: Paint, text: String, cx: Float, cy: Float, textBounds: Rect = Rect()) {
        paint.getTextBounds(text, 0, text.length, textBounds)
        canvas.drawText(text, cx - textBounds.exactCenterX(), cy - textBounds.exactCenterY(), paint)
    }

    companion object {

        private val GAP_ANGLE_DEFAULT = 0f
        private val STROKE_WIDTH_DEFAULT = 3

        private fun dipToPixels(context: Context, dipValue: Float): Float {
            val metrics = context.resources.displayMetrics
            return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dipValue, metrics)
        }
    }
}
